internal/fdo.go

```diff
@@ -64,9 +64,57 @@ func fdoLookupApplication(theme string, size int, appName string) desktop.IconDa
 			return newFdoIconData(theme, size, testLocation)
 		}
 	}
+	//No Match Was Found Checking File Names - Check Files Themselves
+	for _, dataDir := range locationLookup {
+		testLocation := filepath.Join(dataDir, "applications")
+		files, err := ioutil.ReadDir(testLocation)
+		if err != nil {
+			continue
+		}
+		for _, f := range files {
+			if strings.HasPrefix(f.Name(), ".") || f.IsDir() {
+				continue
+			}
+			icon := newFdoIconData(theme, size, filepath.Join(testLocation, f.Name()))
+			if icon != nil {
+				if icon.name == appName {
+					return icon
+				} else if icon.exec == appName {
+					return icon
+				}
+			}
+		}
+	}
 	return nil
 }
 
+//fdoLookupApplicationPartial looks up an application by a partial name and returns all matches in an fdoIconData struct slice
+func fdoLookupApplicationPartialMatch(theme string, size int, appName string) []desktop.IconData {
+	var icons []desktop.IconData
+	locationLookup := fdoLookupXdgDataDirs()
+	for _, dataDir := range locationLookup {
```

Comment:
This chunk of code is almost a compete duplicate of the section above.
A helper function such as "runOnEachApplicationFile"? this would make these functions shorter, reduce the duplication and make it clearer of the purpose.

Reply:
Hmmm... I think on the surface they seem like almost a complete duplicate - but in reality they aren't at all.  The section above stops the loop as soon as it finds an icon and it returns a single icon.  It also looks for exact matches.  This loop does not stop the loop ever and appends every match that is non exact to a slice and then returns.  So in reality one is runOnEachApplicationFileUntilMatched and the other is runOnEachApplicationFile.

Because the loops don't behave the same - one exits early the other must continue - and because the logic within the loops are different - one checks for only one exact match, and the other for multiple partial matches, and that the return values are of different types as well... I'm finding it hard to think of a clean way to separate.